src/main/java/org/team1540/robot2025/subsystems/drive/ModuleIOTalonFX.java

```diff
@@ -0,0 +1,255 @@
+
+    // Turn motor control requests
+    private final VoltageOut turnVoltageReq = new VoltageOut(0);
+    private final TorqueCurrentFOC turnTorqueReq = new TorqueCurrentFOC(0);
+    private final PositionVoltage turnPositionVoltageReq = new PositionVoltage(0.0);
+    private final PositionTorqueCurrentFOC turnPositionTorqueReq = new PositionTorqueCurrentFOC(0.0);
+
+    // Timestamp inputs from Phoenix thread
+    private final Queue<Double> timestampQueue;
+
+    // Inputs from drive motor
+    private final StatusSignal<Angle> drivePosition;
+    private final Queue<Double> drivePositionQueue;
+    private final StatusSignal<AngularVelocity> driveVelocity;
+    private final StatusSignal<Voltage> driveAppliedVolts;
+    private final StatusSignal<Current> driveSupplyCurrent;
+    private final StatusSignal<Current> driveStatorCurrent;
+    private final StatusSignal<Temperature> driveTemp;
+
+    // Inputs from turn motor
+    private final StatusSignal<Angle> turnAbsolutePosition;
+    private final StatusSignal<Angle> turnPosition;
+    private final Queue<Double> turnPositionQueue;
+    private final StatusSignal<AngularVelocity> turnVelocity;
+    private final StatusSignal<Voltage> turnAppliedVolts;
+    private final StatusSignal<Current> turnSupplyCurrent;
+    private final StatusSignal<Current> turnStatorCurrent;
+    private final StatusSignal<Temperature> turnTemp;
+
+    // Connection debouncers
+    private final Debouncer driveConnectedDebounce = new Debouncer(0.5);
+    private final Debouncer turnConnectedDebounce = new Debouncer(0.5);
+    private final Debouncer turnEncoderConnectedDebounce = new Debouncer(0.5);
+
+    public ModuleIOTalonFX(
+            SwerveModuleConstants<TalonFXConfiguration, TalonFXConfiguration, CANcoderConfiguration> constants) {
+        this.constants = constants;
+        ModuleHWConfigs hw = ModuleHWConfigs.fromModuleConstants(constants);
+        drive = new TalonFX(constants.DriveMotorId, DrivetrainConstants.CAN_BUS);
+        turn = new TalonFX(constants.SteerMotorId, DrivetrainConstants.CAN_BUS);
+        cancoder = new CANcoder(constants.EncoderId, DrivetrainConstants.CAN_BUS);
+
+        driveConfig = hw.driveConfig();
+        turnConfig = hw.turnConfig();
+        cancoderConfig = hw.turnEncoderConfig();
+
+        lastDriveConfigStatus =
+                PhoenixUtil.tryUntilOk(5, () -> drive.getConfigurator().apply(driveConfig));
+        lastTurnConfigStatus =
+                PhoenixUtil.tryUntilOk(5, () -> turn.getConfigurator().apply(turnConfig));
+        lastCancoderConfigStatus =
+                PhoenixUtil.tryUntilOk(5, () -> cancoder.getConfigurator().apply(cancoderConfig));
+
+        // Create timestamp queue
+        timestampQueue = OdometryThread.getInstance().makeTimestampQueue();
+
+        // Create drive status signals
+        drivePosition = drive.getPosition();
+        drivePositionQueue = OdometryThread.getInstance().registerSignal(drive.getPosition());
```

Comment:
what is this even doing explain the thread and queues to me

Reply:
The odometry thread is there to take advantage of high-frequency odometry on the CAN FD bus, so that no position value updates from the motors are missed. The odometry thread caches every position update for the motors and gyros in the queues, which are then read off by the IOs every RIO loop cycle and are applied as odometry updates.